curvefs/src/mds/fs_storage.cpp

```diff
@@ -113,5 +124,218 @@ bool MemoryFsStorage::Exist(const std::string &fsName) {
     return true;
 }
 
+uint64_t MemoryFsStorage::NextFsId() {
+    return id_.fetch_add(1, std::memory_order_relaxed);
+}
+
+PersisKVStorage::PersisKVStorage(
+    const std::shared_ptr<curve::kvstorage::KVStorageClient>& storage)
+    : storage_(storage),
+      idGen_(new FsIdGenerator(storage_)),
+      fsLock_(),
+      fs_(),
+      idToNameLock_(),
+      idToName_() {}
+
+PersisKVStorage::~PersisKVStorage() = default;
+
+FSStatusCode PersisKVStorage::Get(uint64_t fsId, FsInfoWrapper* fsInfo) {
+    std::string name;
+    if (!FsIDToName(fsId, &name)) {
+        return FSStatusCode::NOT_FOUND;
+    }
+
+    return Get(name, fsInfo);
+}
+
+bool PersisKVStorage::Init() {
+    bool ret = LoadAllFs();
+    return ret;
+}
+
+void PersisKVStorage::Uninit() {}
+
+FSStatusCode PersisKVStorage::Get(const std::string& fsName,
+                                  FsInfoWrapper* fsInfo) {
+    ReadLockGuard lock(fsLock_);
+    auto iter = fs_.find(fsName);
+    if (iter != fs_.end()) {
+        *fsInfo = iter->second;
+        return FSStatusCode::OK;
+    }
+
+    return FSStatusCode::NOT_FOUND;
+}
+
+FSStatusCode PersisKVStorage::Insert(const FsInfoWrapper& fs) {
+    WriteLockGuard idLock(idToNameLock_);
+    WriteLockGuard fsLock(fsLock_);
+
+    // check if fsname already exists
+    if (fs_.count(fs.FsName()) != 0) {
+        LOG(ERROR) << "fsname already exists, fsname: " << fs.FsName();
+        return FSStatusCode::FS_EXIST;
+    }
+
+    // persist to storage
+    if (!PersitToStorage(fs)) {
+        return FSStatusCode::STORAGE_ERROR;
+    }
+
+    // update cache
+    fs_.emplace(fs.FsName(), fs);
+    idToName_.emplace(fs.FsId(), fs.FsName());
+
+    return FSStatusCode::OK;
+}
+
+FSStatusCode PersisKVStorage::Update(const FsInfoWrapper& fs) {
+    WriteLockGuard lock(fsLock_);
+    auto iter = fs_.find(fs.FsName());
+    if (iter == fs_.end()) {
+        LOG(ERROR) << "fsname not found, fsName: " << fs.FsName();
+        return FSStatusCode::NOT_FOUND;
+    }
+
+    if (iter->second.FsId() != fs.FsId()) {
+        LOG(ERROR) << "fs id not match, fs id in cache: " << iter->second.FsId()
+                   << ", current fs id : " << fs.FsId()
+                   << ", fsName: " << fs.FsName();
+        return FSStatusCode::FS_ID_MISMATCH;
+    }
+
+    // merge with cached info
+    FsInfoWrapper merged = Merge(fs, iter->second);
```

Comment:
It was locked outside in function MountFs and UmountFs，the merge operator is not needed.